samples/connectivity/audio/src/main/java/com/example/platform/connectivity/audio/datasource/AudioLoopSource.kt

```diff
@@ -0,0 +1,160 @@
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+
+/**
+ * Audio Looping, uses AudioRecord and Audio track to loop audio from microphone back to output device
+ * Used for testing microphones and speakers
+ */
+class AudioLoopSource {
+
+    //Scope used for getting buffer from Audio recorder to audio track
+    val coroutineScope = CoroutineScope(Dispatchers.IO)
+    var job: Job? = null
+    var isRecording = MutableStateFlow(false)
+
+    companion object {
+        val sampleRate = 48000
+        val bufferSize = AudioRecord.getMinBufferSize(
+            sampleRate,
+            AudioFormat.CHANNEL_IN_MONO,
+            AudioFormat.ENCODING_PCM_16BIT,
+        )
+
+        val audioTrackBufferSize = AudioTrack.getMinBufferSize(
+            sampleRate,
+            AudioFormat.CHANNEL_OUT_MONO,
+            AudioFormat.ENCODING_PCM_16BIT
+        )
+
+
+        @SuppressLint("MissingPermission")
+        val audioSampler = AudioRecord(
+            MediaRecorder.AudioSource.VOICE_COMMUNICATION,
+            sampleRate,
+            AudioFormat.CHANNEL_IN_MONO,
+            AudioFormat.ENCODING_PCM_16BIT,
+            bufferSize
+        )
+
+        //Audio track for audio playback
+        private var audioTrack = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
+        {
+            AudioTrack.Builder()
+                .setAudioAttributes(
+                    AudioAttributes.Builder()
+                        .setUsage(AudioAttributes.USAGE_VOICE_COMMUNICATION)
+                        .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
+                        .build()
+                )
+                .setAudioFormat(
+                    AudioFormat.Builder()
+                        .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
+                        .setSampleRate(sampleRate)
+                        .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
+                        .build()
+                )
+                .setBufferSizeInBytes(audioTrackBufferSize)
+                .build()
+        }
+        else
+        {
+            AudioTrack(
+                AudioManager.STREAM_VOICE_CALL,
+                sampleRate,
+                AudioFormat.CHANNEL_OUT_MONO,
+                AudioFormat.ENCODING_PCM_16BIT,
+                audioTrackBufferSize,
+                AudioTrack.MODE_STREAM
+            )
+        }
+
+    }
+
+    /**
+     * Gets buffer from Audio Recorder and loops back to the audio track
+     */
+    fun startAudioLoop(): Boolean {
+        if(audioSampler.recordingState == AudioRecord.RECORDSTATE_RECORDING){
+            return false
+        }
+
+        audioTrack.playbackRate = sampleRate
+
+        job = coroutineScope.launch {
+            if (audioSampler.state == AudioRecord.STATE_INITIALIZED) {
+                audioSampler.startRecording()
+            }
+
+            val data = ByteArray(bufferSize)
+            audioTrack.play()
+
```

Comment:
You want to get into a state where the input buffer is empty and the output buffer is full.
This will help avoid glitches.

So you should do non-blocking writes until the AudioTrack is full.
Then do non-blocking reads until the AudioRecord is drained.
Then enter your loop.